docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,381 @@
+    │   │   │   ├── generated_defaulters.go
+    │   │   │   └── generated_default_stubs.go
+    │   │   └── validation
+    │   │       ├── generated_validation.go
+    │   │       └── generated_validation_stubs.go
+    │   └── types.go
+    └── v2alpha1
+        ├── install
+        │   ├── conversion
+        │   │   ├── generated_conversion.go
+        │   │   └── manual_conversion_overrides.go
+        │   ├── defaulting
+        │   │   ├── generated_defaulters.go
+        │   │   └── generated_default_stubs.go
+        │   └── validation
+        │       ├── generated_validation.go
+        │       └── generated_validation_stubs.go
+        └── types.go
+```
+
+### Package: `apis/`
+
+Multiple groups may live in the same tree. Each group gets its own directory
+under `apis/`. We'll assume an `apis/` directory of this structure in all tools.
+
+TBD: Either test code will live here, or no code.
+
+### Package: `apis/group/`
+
+Multiple versions live in a group. Each version gets its own subdirectory.
+
+Additionally, a types.go file contains an "internal" version of the group's
+objects. This file is used as a destination/source for conversion functions.
+
+### Package: `apis/group/install`
+
+This package installs *all* of the versions of the group.
+
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+`types.go` contains the types for this version of the api group.
+
+This package has tightly controlled imports. It's allowed to include the
+Kubernetes unversioned API types (to be moved to `pkg/apis/unversioned`) and
+little else. In particular, it *must not* import its parent `apis/group`.
+
+### Package: `apis/group/v1/install`
+
+This package, if imported, registers the v1 version of this group, its
+conversion, validation, and defaulting functions.
+
+We have separate install packages for every version to allow deliberate import
+choices to be made.
+
+### Package: `apis/group/v1/install/conversion`
+
+This package contains conversion functions, and if imported, will register them.
+
+There are two sorts of conversion functions: those which could be automatically
+generated, and those which have manual additions. All the conversion functions
+convert to & from the types in the parent `apis/groups` and `apis/groups/v1`
+directories.
+
+### Package: `apis/group/v1/install/defaulting`
+
+This package contains defaulting functions, and if imported, will register them.
+
+The functions apply to only the `group/v1` types, *not* the parent `group/`
+types.
+
+There are two sorts of defaulting functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for `// +default=<value>` comments
+preceding the types or their embedded struct members. Similarly, a
+`// +default=custom` marker will force a stub to be generated. Stubs will be
+generated with a `panic("write me!")` marker to ensure that they are populated.
```

Comment:
See https://github.com/kubernetes/kubernetes/pull/16078/files; defaulting needs a way to say "bail out! don't do any more defaulting on this object!"